GPU/Common/DepthRaster.cpp

```diff
@@ -0,0 +1,479 @@
+#include <algorithm>
+#include <cstring>
+
+#include "Common/Math/CrossSIMD.h"
+#include "GPU/Common/DepthRaster.h"
+#include "GPU/Math3D.h"
+#include "Common/Math/math_util.h"
+#include "GPU/Common/VertexDecoderCommon.h"
+
+#if PPSSPP_ARCH(SSE2)
+
+struct Vec4S32 {
+	__m128i v;
+
+	Vec4S32 operator +(Vec4S32 other) const {
+		return Vec4S32{ _mm_add_epi32(v, other.v) };
+	}
+	Vec4S32 operator -(Vec4S32 other) const {
+		return Vec4S32{ _mm_sub_epi32(v, other.v) };
+	}
+	// NOTE: This uses a CrossSIMD wrapper if we don't compile with SSE4 support, and is thus slow.
+	Vec4S32 operator *(Vec4S32 other) const {
+		return Vec4S32{ _mm_mullo_epi32(v, other.v) };   // (ab3,ab2,ab1,ab0)
+	}
+};
+
+struct Vec4F32 {
+	__m128 v;
+
+	static Vec4F32 FromVec4S32(Vec4S32 other) {
+		return Vec4F32{ _mm_cvtepi32_ps(other.v) };
+	}
+
+	Vec4F32 operator +(Vec4F32 other) const {
+		return Vec4F32{ _mm_add_ps(v, other.v) };
+	}
+	Vec4F32 operator -(Vec4F32 other) const {
+		return Vec4F32{ _mm_sub_ps(v, other.v) };
+	}
+	Vec4F32 operator *(Vec4F32 other) const {
+		return Vec4F32{ _mm_mul_ps(v, other.v) };
+	}
+};
+
+#elif PPSSPP_ARCH(ARM_NEON)
+
+struct Vec4S32 {
+	uint32x4_t v;
+
+	Vec4S32 operator +(Vec4S32 other) const {
+		return Vec4S32{ vaddq_s32(v, other.v) };
+	}
+	Vec4S32 operator -(Vec4S32 other) const {
+		return Vec4S32{ vsubq_s32(v, other.v) };
+	}
+	Vec4S32 operator *(Vec4S32 other) const {
+		return Vec4S32{ vmulq_s32(v, other.v) };
+	}
+};
+
+struct Vec4F32 {
+	float32x4_t v;
+
+	static Vec4F32 FromVec4S32(Vec4S32 other) {
+		return Vec4F32{ _mm_cvtepi32_ps(other.v) };
+	}
+
+	Vec4F32 operator +(Vec4F32 other) const {
+		return Vec4F32{ vaddq_f32(v, other.v) };
+	}
+	Vec4F32 operator -(Vec4F32 other) const {
+		return Vec4F32{ vsubq_f32(v, other.v) };
+	}
+	Vec4F32 operator *(Vec4F32 other) const {
+		return Vec4F32{ vmulq_f32(v, other.v) };
+	}
+};
+
+#else
+
+struct Vec4S32 {
+	s32 v[4];
+};
+
+#endif
+
+struct ScreenVert {
+	int x;
+	int y;
+	uint16_t z;
+	uint16_t behind;
+};
+
+void DepthRasterRect(uint16_t *dest, int stride, int x1, int y1, int x2, int y2, short depthValue, GEComparison depthCompare) {
+	// Swap coordinates if needed, we don't back-face-cull rects.
+	// We also ignore the UV rotation here.
+	if (x1 > x2) {
+		std::swap(x1, x2);
+	}
+	if (y1 > y2) {
+		std::swap(y1, y2);
+	}
+	if (x1 == x2 || y1 == y2) {
+		return;
+	}
+
+#if PPSSPP_ARCH(SSE2)
+	__m128i valueX8 = _mm_set1_epi16(depthValue);
+	for (int y = y1; y < y2; y++) {
+		__m128i *ptr = (__m128i *)(dest + stride * y + x1);
+		int w = x2 - x1;
+		switch (depthCompare) {
+		case GE_COMP_ALWAYS:
+			while (w >= 8) {
+				_mm_storeu_si128(ptr, valueX8);
+				ptr++;
```

Comment:
Nitpick, but incrementing possibly-misaligned pointer feels a bit iffy (though compilers probably won't do anything weird here), compared to only casting inside `_mm_storeu_si128(...)`.

Reply:
I do think this is fine actually, _mm_storeu_si128 is meant to handle misaligned pointers, unlike _mm_store_si128. (although on modern CPUs, they are the same).